### PR TITLE
[Deprecated] [Fix] Add `avg_non_ignore` in ce loss.

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation
Right now, MMSegmentation do not support `ignore_index` in loss average over non-ignored elements in CE loss.

fix https://github.com/open-mmlab/mmsegmentation/pull/578.

In old PR: https://github.com/open-mmlab/mmsegmentation/pull/578, the author fixed it while it left a bug if all labels in one image is `ignore_index`, `avg_factor` would be zero and ZeroDivisionError would be rasied in `weight_reduce_loss` function.

## Usage
Just take [fcn_unet_s5-d16_128x128_40k_chase_db1.py](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/unet/fcn_unet_s5-d16_128x128_40k_chase_db1.py), its [base config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/_base_/models/fcn_unet_s5-d16.py) for example:

original config:
```python
# model settings
...
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
...
```
image

Just add `ignore_index` in decoder head or auxiliary head and add `avg_non_ignore=True`:
```python
# model settings
...
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0, avg_non_ignore=True),
...
```

image

|  cfg   | original  |  `avg_non_ignore=True`  |  `avg_non_ignore=False`  |
|  ----  | ----  | ----  | ----  |
| deeplabv3_r18-d8_512x1024_80k_cityscapes  | 76.70 78.27(ms+flip) | 77.07 78.82(ms+flip) | - |
| deeplabv3_r50-d8_512x512_80k_ade20k | 42.42 43.28(ms+flip) | 43.08 44.31(ms+flip) | 42.40 43.65(ms+flip) |

### CODE REVIEW

File: mmseg/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -13,7 +13,8 @@ def cross_entropy(pred,
                   class_weight=None,
                   reduction='mean',
                   avg_factor=None,
-                  ignore_index=-100):
+                  ignore_index=-100,
+                  ignore_index_flag=False):
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:17]: ignore_index_flag -> avg_non_ignore

File: mmseg/models/losses/utils.py [hunk 2]
```diff
@@ -69,7 +69,9 @@ def weight_reduce_loss(loss, weight=None, reduction='mean', avg_factor=None):
     else:
         # if reduction is mean, then average the loss by avg_factor
         if reduction == 'mean':
-            loss = loss.sum() / avg_factor
+            # Avoid causing ZeroDivisionError when avg_factor is 0.0,
+            # i.e., all labels of an image belong to ignore index.
+            loss = loss.sum() / avg_factor if avg_factor != 0.0 else 0.0
```
Comment (MeowZheng) [inline mmseg/models/losses/utils.py:74]: avg_factor + eps

File: mmseg/models/losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -85,6 +94,16 @@ def binary_cross_entropy(pred,
...
+    elif ignore_index_flag:
+        # should mask out the ignored elements
+        valid_mask = ((label >= 0) & (label != ignore_index)).float()
+        if weight is not None:
+            weight *= valid_mask
+        else:
+            weight = valid_mask
+        # average loss over non-ignored elements
+        if avg_factor is None:
+            avg_factor = label.numel() - (label == ignore_index).sum().item()
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:106]: I think ignore index is not necessary for binary classification

File: mmseg/models/losses/cross_entropy_loss.py [hunk 4]
```diff
@@ -184,6 +210,7 @@ def forward(self,
                 weight=None,
                 avg_factor=None,
                 reduction_override=None,
+                ignore_index=None,
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:213]: kwargs includes ignore_index

File: mmseg/models/losses/cross_entropy_loss.py [hunk 5]
```diff
@@ -13,7 +13,8 @@ def cross_entropy(pred,
                   class_weight=None,
                   reduction='mean',
                   avg_factor=None,
-                  ignore_index=-100):
+                  ignore_index=-100,
+                  avg_non_ignore=False):
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:17]: might add docstring

File: mmseg/models/losses/utils.py [hunk 6]
```diff
@@ -69,7 +70,11 @@ def weight_reduce_loss(loss, weight=None, reduction='mean', avg_factor=None):
     else:
         # if reduction is mean, then average the loss by avg_factor
         if reduction == 'mean':
-            loss = loss.sum() / avg_factor
+            # Avoid causing ZeroDivisionError when avg_factor is 0.0,
+            # i.e., all labels of an image belong to ignore index.
+            eps = torch.finfo(torch.float32).eps
+            loss = loss.sum() / (avg_factor +
+                                 eps) if avg_factor != 0.0 else 0.0
```
Comment (MeowZheng) [inline mmseg/models/losses/utils.py:77]:
```suggestion
            loss = loss.sum() / avg_factor + eps
```

**Review by MeowZheng — COMMENTED** [review]
might add some ut for avg_non_ignore

File: mmseg/models/losses/cross_entropy_loss.py [hunk 7]
```diff
@@ -25,6 +26,11 @@ def cross_entropy(pred,
         ignore_index=ignore_index)
 
     # apply weights and do the reduction
+    # average loss over non-ignored elements
+    # pytorch's official cross_entropy average loss over non-ignored elements
+    # refer to https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/nn/functional.py#L2660  # noqa
+    if (avg_factor is None) and avg_non_ignore:
+        avg_factor = label.numel() - (label == ignore_index).sum().item()
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:33]: loss might be masked by ignore_index

File: mmseg/models/losses/utils.py [hunk 8]
```diff
@@ -69,7 +70,10 @@ def weight_reduce_loss(loss, weight=None, reduction='mean', avg_factor=None):
     else:
         # if reduction is mean, then average the loss by avg_factor
         if reduction == 'mean':
-            loss = loss.sum() / avg_factor
+            # Avoid causing ZeroDivisionError when avg_factor is 0.0,
+            # i.e., all labels of an image belong to ignore index.
+            eps = torch.finfo(torch.float32).eps
+            loss = loss.sum() / (avg_factor + eps)
```
Comment (MeowZheng) [inline mmseg/models/losses/utils.py:76]: loss in ignored pixels might not be summed

File: mmseg/models/losses/cross_entropy_loss.py [hunk 9]
```diff
@@ -13,7 +13,26 @@ def cross_entropy(pred,
...
+        class_weight (list[float], optional): The weight for each class.
+            Default: None.
+        reduction (str, optional): The method used to reduce the loss.
+            Options are "none", "mean" and "sum". Default: 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Default: None.
+        ignore_index (int): Index that will be ignored in evaluation.
+            Default: -100.
+        avg_non_ignore (bool, optional): Average factor whether to ignore index
+            when computing the mean of losses. Default: False.
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:34]:
```suggestion
        avg_non_ignore (bool): The flag decides to whether the loss is only averaged over 
            non-ignored targets. Default: False.
```

File: mmseg/models/losses/cross_entropy_loss.py [hunk 10]
```diff
@@ -13,7 +13,26 @@ def cross_entropy(pred,
...
+        weight (torch.Tensor, optional): Sample-wise loss weight.
+            Default: None.
+        class_weight (list[float], optional): The weight for each class.
+            Default: None.
+        reduction (str, optional): The method used to reduce the loss.
+            Options are "none", "mean" and "sum". Default: 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Default: None.
+        ignore_index (int): Index that will be ignored in evaluation.
+            Default: -100.
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:32]:
```suggestion
        ignore_index (int): Specifies a target value that is ignored and does not contribute to the input gradients.
            When ``avg_non_ignore `` is ``True``, and the ``reduction`` is ``''mean''``, 
            the loss is averaged over non-ignored targets. Defaults: -100.
```

File: tests/test_models/test_losses/test_ce_loss.py [hunk 11]
```diff
@@ -86,4 +86,78 @@ def test_ce_loss():
...
+        loss_weight=1.0,
+        avg_non_ignore=True)
+    loss_cls = build_loss(loss_cls_cfg)
+    loss = loss_cls(fake_pred, fake_label, ignore_index=10)
+    torch_loss = torch.nn.functional.cross_entropy(
+        fake_pred,
+        fake_label,
+        ignore_index=10,
+        reduction='sum',
+        weight=class_weight) / 12.0
```
Comment (MeowZheng) [inline tests/test_models/test_losses/test_ce_loss.py:131]: why not use reduction = 'mean'?

**Comment by ZwwWayne** [comment]
Suggest @RangiLyu take a look. Related PR in mmdet: https://github.com/open-mmlab/mmdetection/pull/7449.
Hopefully, we can have consistent usages in both mmseg & mmdet.

File: tests/test_models/test_losses/test_ce_loss.py (hunk 11 above)
Comment (MengzhangLI) [reply tests/test_models/test_losses/test_ce_loss.py:131]: It is then divided by `12.0`, which is the number of non-ignore-index pixels in `fake_label`.

**Comment by MengzhangLI** [comment]
Test again:

Formal training: 
image

with `avg_non_ignore` and `ignore_index=0`:
image

File: mmseg/models/losses/cross_entropy_loss.py [hunk 12]
```diff
@@ -161,14 +204,16 @@ def __init__(self,
                  reduction='mean',
                  class_weight=None,
                  loss_weight=1.0,
-                 loss_name='loss_ce'):
+                 loss_name='loss_ce',
+                 avg_non_ignore=False):
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:208]: might add a warning when `avg_non_ignore` is False to tell users if they would like to ignore some pixels they must open `avg_non_ignore` manually, `avg_non_ignore` defaults to closed

File: tests/test_models/test_losses/test_ce_loss.py (hunk 11 above)
Comment (MeowZheng) [reply tests/test_models/test_losses/test_ce_loss.py:131]: but ce from torch, just average loss on the non-ignore targets, when reduction is mean

File: mmseg/models/losses/cross_entropy_loss.py [hunk 13]
```diff
@@ -85,6 +117,16 @@ def binary_cross_entropy(pred,
             'H, W], label shape [N, H, W] are supported'
         label, weight = _expand_onehot_labels(label, weight, pred.shape,
                                               ignore_index)
+    elif avg_non_ignore:
+        # should mask out the ignored elements
+        valid_mask = ((label >= 0) & (label != ignore_index)).float()
+        if weight is not None:
+            weight *= valid_mask
+        else:
+            weight = valid_mask
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:126]: weight returned from `_expand_onehot_labels` has been masked

File: mmseg/models/losses/cross_entropy_loss.py [hunk 14]
```diff
@@ -161,14 +206,22 @@ def __init__(self,
...
+                 avg_non_ignore=False):
         super(CrossEntropyLoss, self).__init__()
         assert (use_sigmoid is False) or (use_mask is False)
         self.use_sigmoid = use_sigmoid
         self.use_mask = use_mask
         self.reduction = reduction
         self.loss_weight = loss_weight
         self.class_weight = get_class_weight(class_weight)
+        self.avg_non_ignore = avg_non_ignore
+        if not self.avg_non_ignore:
```
Comment (xvjiarui) [inline mmseg/models/losses/cross_entropy_loss.py:219]: This warning is a little bit confusing. 
We should warn users when `avg_non_ignore = True`, the loss will be slight larger blablabla...

Comment (MeowZheng) [reply mmseg/models/losses/cross_entropy_loss.py:219]: Some users prefer `avg_non_ignore` defaults to `True`, as they think there might be a bug in the original loss calculation. This warning works for reminding them to open `avg_non_ignore`.

I think msg should be refined to ba more straightforwardly.

File: mmseg/models/losses/cross_entropy_loss.py [hunk 15]
```diff
@@ -161,14 +206,22 @@ def __init__(self,
...
+            warnings.warn(
+                'Default ``avg_non_ignore`` is False, if you would like to '
+                'ignore certain label and average loss over non-ignore label, '
+                'open ``avg_non_ignore=True`` and set ``ignore_index`` '
+                'in decoder or auxiliary heads config file')
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:224]:
```suggestion
                'Default ``avg_non_ignore`` is False, if you would like to '
                'ignore certain label and average loss over non-ignore label, '
                'set ``avg_non_ignore=True``')
```

File: mmseg/models/losses/cross_entropy_loss.py (hunk 14 above)
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:219]:
```suggestion
        if not self.avg_non_ignore and reduction=='mean', :
```

File: mmseg/models/losses/cross_entropy_loss.py [hunk 16]
```diff
@@ -161,14 +206,22 @@ def __init__(self,
...
+                 loss_name='loss_ce',
+                 avg_non_ignore=False):
         super(CrossEntropyLoss, self).__init__()
         assert (use_sigmoid is False) or (use_mask is False)
         self.use_sigmoid = use_sigmoid
         self.use_mask = use_mask
         self.reduction = reduction
         self.loss_weight = loss_weight
         self.class_weight = get_class_weight(class_weight)
+        self.avg_non_ignore = avg_non_ignore
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:218]: add `avg_non_ignore` in `extra_repr`

File: mmseg/models/losses/cross_entropy_loss.py (hunk 15 above)
Comment (MengzhangLI) [reply mmseg/models/losses/cross_entropy_loss.py:224]: May `ignore_index` also be reminded to set in warnings thus users would not leave it out?

Comment (MengzhangLI) [reply mmseg/models/losses/cross_entropy_loss.py:224]: In config file, `ignore_index` belongs to decoder/auxiliary head but `avg_non_ignore` belongs to loss_decode.

File: mmseg/models/losses/cross_entropy_loss.py (hunk 13 above)
Comment (MengzhangLI) [reply mmseg/models/losses/cross_entropy_loss.py:126]: `_expand_onehot_labels` only worked when `pred.dim() != label.dim()`. But in the case when `pred.dim() == label.dim()`, it would not be called.

File: mmseg/models/losses/cross_entropy_loss.py [hunk 17]
```diff
@@ -85,6 +117,16 @@ def binary_cross_entropy(pred,
...
+    elif avg_non_ignore:
+        # should mask out the ignored elements
+        valid_mask = ((label >= 0) & (label != ignore_index)).float()
+        if weight is not None:
+            weight *= valid_mask
+        else:
+            weight = valid_mask
+        # average loss over non-ignored elements
+        if avg_factor is None:
+            avg_factor = label.numel() - (label == ignore_index).sum().item()
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:129]:
```suggestion
    else:
        # should mask out the ignored elements
        valid_mask = ((label >= 0) & (label != ignore_index)).float()
        if weight is not None:
            weight *= valid_mask
        else:
            weight = valid_mask
   # average loss over non-ignored elements
   if avg_factor is None:
        avg_factor = label.numel() - (label == ignore_index).sum().item()
```

**Comment by MengzhangLI** [comment]
Due to a lot of unexpected and unrelated files to be merged in this PR, it would be closed and a new pr would be created.

**Comment by MengzhangLI** [comment]
> Suggest @RangiLyu take a look. Related PR in mmdet: [open-mmlab/mmdetection#7449](https://github.com/open-mmlab/mmdetection/pull/7449). Hopefully, we can have consistent usages in both mmseg & mmdet.

https://github.com/open-mmlab/mmsegmentation/pull/1409 new pr is here.
